### PR TITLE
 Add the "type" URL paramenter to sendUpdateValueNgsi2

### PR DESCRIPTION
**Pull request by danvilmot**
Add the ULR parameter "type" to specify the type in case of ambiguity when there are several entities with the same "id" but different types as described in the issue #733

### CODE REVIEW

File: lib/services/ngsi/ngsiService.js [hunk 1]
```diff
@@ -428,7 +428,7 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
     var payload = {};
 
     var options = createRequestObject(
-        '/v2/entities/' + entityName + '/attrs',
+        '/v2/entities/' + entityName + '/attrs' + '?type=' + typeInformation.type,
```
Comment (AlvaroVega) [inline lib/services/ngsi/ngsiService.js]: Maybe check if type is in typeInformation?

Comment (danvilmot) [reply lib/services/ngsi/ngsiService.js]: Do you mean, for example, initialize the var ``options`` after the ``if`` in https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js#L439 ?

File: lib/services/ngsi/ngsiService.js [hunk 2]
```diff
@@ -426,11 +426,20 @@ function castJsonNativeAttributes(payload) {
 function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, callback) {
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]: I wonder if some other request appart from sending updates to CB could potencially have the same problem described in #733. Maybe looking for `/v2/entities` in the code could provide some hints.

What do you think @dcalvoalonso?

**Comment by fgalan** [comment]
@danvilmot thanks for the contribution! Please find some inline comments in the code.

In addition, note that travis CI is failing. It seems that the code have some linting problems. Please have a look to travis report for additional information.

File: lib/services/ngsi/ngsiService.js (hunk 2 above)
Comment (dcalvoalonso) [reply lib/services/ngsi/ngsiService.js]: Hi @fgalan, it is possible... I will review this.

**Comment by fgalan** [comment]
It seems that 63 unit test are failing. Maybe this is due the change of expectations in some mock.

File: lib/services/ngsi/ngsiService.js (hunk 2 above)
Comment (fgalan) [reply lib/services/ngsi/ngsiService.js]:
```
$ grep -R '/v2/entities' lib/
lib/services/ngsi/ngsiService.js:        '/v2/entities/' + entityName + '/attrs',
lib/services/ngsi/ngsiService.js:    var url = '/v2/entities/' + entityName + '/attrs';
lib/services/devices/deviceService.js:        url: config.getConfig().contextBroker.url + '/v2/entities?options=upsert',
lib/services/devices/deviceService.js:        options.url = deviceData.cbHost + '/v2/entities?options=upsert';
lib/services/devices/deviceService.js:        options.url = 'http://' + deviceData.cbHost + '/v2/entities?options=upsert';
lib/services/devices/deviceService.js:        url: config.getConfig().contextBroker.url + '/v2/entities/' + String(deviceData.name) + '/attrs',
lib/services/devices/deviceService.js:        options.url = deviceData.cbHost + '/v2/entities/' + String(deviceData.name) + '/attrs';
lib/services/devices/deviceService.js:        options.url = 'http://' + deviceData.cbHost + '/v2/entities/' + String(deviceData.name) + '/attrs';
````
Now I'll have a closer look to each one...

File: lib/services/ngsi/ngsiService.js [hunk 3]
```diff
@@ -426,11 +426,22 @@ function castJsonNativeAttributes(payload) {
...
 
     var payload = {};
-
-    var options = createRequestObject(
-        '/v2/entities/' + entityName + '/attrs',
-        typeInformation,
-        token);
+
```
Comment (fgalan) [inline lib/services/ngsi/ngsiService.js]: I'd suggest the following alternative (more compact):

```
    var url = '/v2/entities/' + entityName + '/attrs';
    if (typeInformation.type) {
       url += '?type=' + typeInformation.type;
    } 
   var options = createRequestObject(url, typeInformation, token);
```

File: lib/services/ngsi/ngsiService.js (hunk 2 above)
Comment (fgalan) [reply lib/services/ngsi/ngsiService.js]: In ngsiService.js the first one correspond to sendUpdateValueNgsi2() already included in this PR. The second one corresponds to sendQueryValueNgsi2() and I think that `?type` should be also added there.

In deviceService.js the three first occurences are in createInitialEntityNgsi2() so they don't need `?type` (entity type is already included in the request payload in this case). The three last ocurrences are in updateEntityNgsi2() and probably that one could also have ambiguity problems so `?type` should be included.

@dcalvoalonso what do you think? Do you agree with my analysis or is something wrong?

**Comment by fgalan** [comment]
With regards to test fails, it seems it is due to fail in mock expectations. Fortunatelly, this is easy to fix.

You will get fragments like this one when running `npm test`:

```
  82) Timestamp processing plugin
       When an update comes with a timestamp through the plugin
         should return an entity with all its timestamps expanded to have separators:
     Uncaught AssertionError: expected Error {
  status: 404,
  statusCode: 404,
  message: 'Nock: No match for request {\n  "method": "POST",\n  "url": "http://192.168.1.1:1026/v2/entities/light1/attrs?type=Light",\n  "headers": {\n    "fiware-service": "smartGondor",\n    "fiware-servicepath": "gardens",\n    "host": "192.168.1.1:1026",\n    "accept": "application/json",\n    "content-type": "application/json",\n    "content-length": 192\n  },\n  "body": "{\\"state\\":{\\"type\\":\\"Boolean\\",\\"value\\":true,\\"metadata\\":{\\"TimeInstant\\":{\\"type\\":\\"DateTime\\",\\"value\\":\\"2016-05-30T16:25:22.304Z\\"}}},\\"TimeInstant\\":{\\"type\\":\\"DateTime\\",\\"value\\":\\"2016-05-30T16:25:22.304Z\\"}}"\n}'
} to not exist
      at /data/src/iotagent-node-lib/test/unit/ngsiv2/plugins/timestamp-processing-plugin_test.js:112:28
      at Request._callback (lib/services/ngsi/ngsiService.js:131:13)
      at self.callback (node_modules/request/request.js:185:22)
      at Request.onRequestError (node_modules/request/request.js:881:8)
      at /data/src/iotagent-node-lib/node_modules/nock/lib/request_overrider.js:221:11
      at _combinedTickCallback (internal/process/next_tick.js:132:7)
      at process._tickDomainCallback (internal/process/next_tick.js:219:9)
```

So you go to the referred line (test/unit/ngsiv2/plugins/timestamp-processing-plugin_test.js:112) and look fo the mock definition which use to be some lines above the referred line. It shows somethink like that:

```
contextBrokerMock = nock('http://192.168.1.1:1026')
                .matchHeader('fiware-service', 'smartGondor')
                .matchHeader('fiware-servicepath', 'gardens')
                .post('/v2/entities/light1/attrs', utils.readExampleFile(
                    './test/unit/ngsiv2/examples/contextRequests/updateContextProcessTimestamp.json'))
                .reply(204);
```

You need to add an expectation for `?type` adding a .query() to the sequence, just after .post()/.get():

```
contextBrokerMock = nock('http://192.168.1.1:1026')
                .matchHeader('fiware-service', 'smartGondor')
                .matchHeader('fiware-servicepath', 'gardens')
                .query({type: 'Light'})
                .post('/v2/entities/light1/attrs', utils.readExampleFile(
                    './test/unit/ngsiv2/examples/contextRequests/updateContextProcessTimestamp.json'))
                .reply(204);
```

Hint: have a look to the error message to check which type is used in each case (e.g. `"url": "http://192.168.1.1:1026/v2/entities/light1/attrs?type=Light"`).

To check it has worked, `npm test` shouldn't show that test failing.

I guess most of the test currently failing (maybe all) can be fixed this way. @danvilmot could you manage to modify the PR in order to do so, please? Thanks!

File: lib/services/ngsi/ngsiService.js (hunk 2 above)
Comment (fgalan) [reply lib/services/ngsi/ngsiService.js]: Great. So let's wait to @danvilmot to finish the work with tests (and maybe this: https://github.com/telefonicaid/iotagent-node-lib/pull/734#discussion_r249471301) then you can continue from that part once this gets merged. Thanks!

NTC

**Comment by fgalan** [comment]
Linting error precules travis executing the tests:

```
test/unit/ngsiv2/general/contextBrokerOAuthSecurityAccess-test.js: line 504, col 135, Line is too long.
```

@danvilmot could you fix it in this PR branch, pls? It's easy ;)

**Comment by fgalan** [comment]
Now travis result is:

```
  458 passing (47s)
  7 pending
  33 failing
```

So we are progresssing (last time it was 63 fails ;) but it is not done yet. Not sure if the 33 failing tests follow the same pattern or there are other causes.

**Comment by danvilmot** [comment]
Still one test fails:
```
1) Expression-based transformations plugin
       When there are expression attributes that are just calculated (not sent by the device)
         should calculate them and add them to the payload:
     Error: Timeout of 3000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/travis/build/telefonicaid/iotagent-node-lib/test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js)
```

I do not how to tackle.

**Comment by fgalan** [comment]
Great! We are progressing :)

I'd have a look to that test and try to provide some insight about how to fix it.

**Comment by fgalan** [comment]
@danvilmot it seems a `.query({type: 'WeatherStation')` is missing just after line https://github.com/telefonicaid/iotagent-node-lib/pull/734/files#diff-d17b42cacba3d04a6a2daa6922eb57b3R260. That should fix the test.

**Comment by dcalvoalonso** [comment]
LGTM

File: lib/services/ngsi/ngsiService.js [hunk 4]
```diff
@@ -426,11 +426,20 @@ function castJsonNativeAttributes(payload) {
...
```
[the remark this answers is not in the document]
Comment (fgalan) [reply lib/services/ngsi/ngsiService.js]: Ups... it seems CNR was not modified at the end. I'll add a comment in #733 for not forgetting this.

**Pull request merged**